src/gpl/src/gpuSolver.cu

```diff
@@ -0,0 +1,208 @@
+
+  // transform from coordinates (COO) values to compressed row pointers (CSR)
+  // values https://docs.nvidia.com/cuda/cusparse/index.html
+  int* r_csrRowInd = NULL;
+  thrust::device_vector<int> d_csrRowInd(m_ + 1, 0);
+  r_csrRowInd = thrust::raw_pointer_cast(d_csrRowInd.data());
+
+  cusparseerror(cusparseXcoo2csr(handleCusparse,
+                                 r_cooRowIndex_,
+                                 nnz_,
+                                 m_,
+                                 r_csrRowInd,
+                                 CUSPARSE_INDEX_BASE_ZERO));
+
+  cusolvererror(cusolverSpScsrlsvqr(handleCusolver,
+                                    m_,
+                                    nnz_,
+                                    descr,
+                                    r_cooVal_,
+                                    r_csrRowInd,
+                                    r_cooColIndex_,
+                                    r_fixedInstForceVec_,
+                                    tol,
+                                    reorder,
+                                    r_instLocVec_,
+                                    &singularity));
+
+  // Sync and Copy data to host
+  cudaerror(cudaMemcpyAsync(instLocVec.data(),
+                            r_instLocVec_,
+                            sizeof(float) * m_,
+                            cudaMemcpyDeviceToHost,
+                            stream));
+
+  // cudaerror(cudaFree(r_csrRowInd));
+  cusparseerror(cusparseDestroyMatDescr(descr));
+  cusparseerror(cusparseDestroy(handleCusparse));
+  cusolvererror(cusolverSpDestroy(handleCusolver));
+}
+
+__global__ void Multi_MatVec(float error,
+                             int nnz_,
+                             int m_,
+                             float* r_Ax,
+                             float* r_fixedInstForceVec_,
+                             float* r_instLocVec_,
+                             int* r_cooRowIndex_,
+                             int* r_cooColIndex_,
+                             float* r_cooVal_)
+{
+  float sum = 0;
+  int num = blockIdx.x * blockDim.x + threadIdx.x;
+  if (num < nnz_) {
+    r_Ax[r_cooRowIndex_[num]]
+        += r_cooVal_[num] * r_instLocVec_[r_cooColIndex_[num]];
+  }
+  for (size_t row = 0; row < m_; row++) {
+    sum += (r_fixedInstForceVec_[row] > 0) ? r_fixedInstForceVec_[row]
+                                           : -r_fixedInstForceVec_[row];
+    if (r_fixedInstForceVec_[row] > r_Ax[row])
+      error += r_fixedInstForceVec_[row] - r_Ax[row];
+    else
+      error -= r_fixedInstForceVec_[row] - r_Ax[row];
+  }
+  if (sum != 0)
+    error = error / sum;
+}
+
+float GpuSolver::error()
+{
+  float* r_Ax;
+  thrust::device_vector<float> t_Ax(m_);
+  thrust::fill(t_Ax.begin(), t_Ax.end(), 0);
+  r_Ax = thrust::raw_pointer_cast(t_Ax.data());
+
+  unsigned int threads = 512;
+  unsigned int blocks = (m_ + threads - 1) / threads;
+  Multi_MatVec<<<blocks, threads>>>(error_,
```

Comment:
are all the <<< needed?

Reply:
Yes.

Reply:
That's the CUDA kernel launch syntax. It is needed